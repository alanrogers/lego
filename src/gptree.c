/**
 * @file gptree.c
 * @brief Methods for simulating gene genealogies within a given tree
 * of populations, and allowing populations to mix and also to split.
 */

#include "gptree.h"
#include "gene.h"
#include "lblndx.h"
#include "parse.h"
#include "parstore.h"
#include <string.h>

struct GPTree {
    int nseg; // number of segments in population tree.
    PopNode *pnv; // array of length nseg
    PopNode *rootPop;
    Gene *rootGene;
    Bounds bnd;
    ParStore *parstore;
    LblNdx lblndx;
    SampNdx sndx;
};

void GPTree_setParams(GPTree *self, int n, double x[n]) {
    assert(n == ParStore_nFree(self->parstore));
    ParStore_setFreeParams(self->parstore, n, x);
}

/// Return number of free parameters
int GPTree_nFree(const GPTree *self) {
    return ParStore_nFree(self->parstore);
}

void GPTree_simulate(GPTree *self, BranchTab *branchtab, gsl_rng *rng,
                     unsigned long nreps) {
    unsigned long rep;
    for(rep = 0; rep < nreps; ++rep) {
        PopNode_clear(self->rootPop); // remove old samples 
        SampNdx_populateTree(&(self->sndx));    // add new samples

        // coalescent simulation generates gene genealogy within
        // population tree.
        self->rootGene = PopNode_coalesce(self->rootPop, rng);
        assert(self->rootGene);

        // Traverse gene tree, accumulating branch lengths in bins
        // that correspond to site patterns.
        Gene_tabulate(self->rootGene, branchtab);

        // Free gene genealogy but not population tree.
        Gene_free(self->rootGene);
        self->rootGene = NULL;
    }
}

GPTree *GPTree_new(const char *fname, Bounds bnd) {
    GPTree *self = malloc(sizeof(GPTree));
    CHECKMEM(self);

    memset(self, 0, sizeof(GPTree));
    self->bnd = bnd;
    self->parstore = ParStore_new();
    LblNdx_init(&self->lblndx);
    SampNdx_init(&self->sndx);
    FILE *fp = fopen(fname, "r");
    if(fp == NULL)
        eprintf("%s:%s:%d: can't open file \"%s\".\n",
                __FILE__, __func__, __LINE__, fname);
    self->nseg = countSegments(fp);
    rewind(fp);
    self->pnv = malloc(self->nseg * sizeof(self->pnv[0]));
    CHECKMEM(self->pnv);

    NodeStore *ns = NodeStore_new(self->nseg, self->pnv);
    CHECKMEM(ns);

    self->rootPop = mktree(fp, &self->sndx, &self->lblndx, self->parstore,
                           &self->bnd, ns);

    fclose(fp);
    NodeStore_free(ns);
    GPTree_sanityCheck(self, __FILE__, __LINE__);
    return self;
}

/// Destructor.
void GPTree_free(GPTree *self) {
    Gene_free(self->rootGene);
    self->rootGene = NULL;
    PopNode_clear(self->rootPop);
    self->rootPop = NULL;
    free(self->pnv);
    ParStore_free(self->parstore);
    free(self);
}

/// Duplicate a GPTree object
GPTree *GPTree_dup(const GPTree *old) {
    if(old->rootGene != NULL)
        eprintf("%s:%s:%d: old->rootGene must be NULL on entry\n",
                __FILE__,__func__,__LINE__);
    if(!PopNode_isClear(old->rootPop))
        eprintf("%s:%s:%d: clear GPTree of samples before call"
                " to GPTree_dup\n",
                __FILE__,__func__,__LINE__);

    GPTree *new   = memdup(old, sizeof(GPTree));
    new->parstore = ParStore_dup(old->parstore);
    new->pnv      = memdup(old->pnv, old->nseg * sizeof(PopNode));
    assert(old->nseg == new->nseg);

    new->sndx = old->sndx;

    /*
     * Adjust the pointers within each PopNode object so they refer to
     * the memory allocated in "new" rather than that in "old".
     * dpar is the offset between new and old for parameter pointers.
     * dpop is the analogous offset for PopNode pointers. Everything
     * has to be cast to size_t, because we are not doing pointer
     * arithmetic in the usual sense. Ordinarily, ptr+3 means
     * ptr + 3*sizeof(*ptr). We want it to mean ptr+3*sizeof(char).
     */
    int i;
    size_t dpar = ((size_t) new->parstore) - ((size_t) old->parstore);
    size_t dpop = ((size_t) new->pnv) - ((size_t) old->pnv);
    INCR_PTR(new->rootPop, dpop);
    for(i=0; i < old->nseg; ++i) {
        PopNode_shiftParamPtrs(&new->pnv[i], dpar);
        PopNode_shiftPopNodePtrs(&new->pnv[i], dpop);
    }
    SampNdx_shiftPtrs(&new->sndx, dpop);
    assert(SampNdx_ptrsLegal(&new->sndx, new->pnv, new->pnv + new->nseg));

    GPTree_sanityCheck(new, __FILE__, __LINE__);
    return new;
}

void GPTree_sanityCheck(GPTree *self, const char *file, int line) {
#ifndef NDEBUG
    REQUIRE(self->nseg > 0,                         file, line);
    REQUIRE(self->pnv != NULL,                      file, line);
    REQUIRE(self->rootPop >= self->pnv,             file, line);
    REQUIRE(self->rootPop < self->pnv + self->nseg, file, line);
    Bounds_sanityCheck(&self->bnd,                  file, line);
    ParStore_sanityCheck(self->parstore,            file, line);
    LblNdx_sanityCheck(&self->lblndx,               file, line);
#endif
}

/// Return 1 if two GPTree objects are equal, 0 if they differ.  Abort
/// with an error if the GPTree pointers are different but one or more
/// of the internal pointers is the same.  Does not access rootPop or
/// rootGene.
int GPTree_equals(GPTree *lhs, GPTree *rhs) {
    if(lhs == rhs)
        return 0;
    if(lhs->pnv == rhs->pnv)
        eprintf("%s:%s:%d: two GPTree objects share a PopNode pointer\n",
                __FILE__,__func__,__LINE__);
    if(lhs->parstore == rhs->parstore)
        eprintf("%s:%s:%d: two GPTree objects share a ParStore pointer\n",
                __FILE__,__func__,__LINE__);
    if(!Bounds_equals(&lhs->bnd, &rhs->bnd))
        return 0;
    if(!ParStore_equals(lhs->parstore, rhs->parstore))
        return 0;
    if(!LblNdx_equals(&lhs->lblndx, &rhs->lblndx))
        return 0;
    if(!SampNdx_equals(&lhs->sndx, &rhs->sndx))
        return 0;
    return 1;
}

LblNdx GPTree_getLblNdx(GPTree *self) {
    return self->lblndx;
}

/// Return pointer to array of lower bounds of free parameters
double     *GPTree_loBounds(GPTree *self) {
    return ParStore_loBounds(self->parstore);
}

/// Return pointer to array of upper bounds of free parameters
double     *GPTree_upBounds(GPTree *self) {
    return ParStore_upBounds(self->parstore);
}

/// Return number of samples.
unsigned    GPTree_nsamples(GPTree *self) {
    return SampNdx_size(&self->sndx);
}


#ifdef TEST

#include <string.h>
#include <assert.h>

#ifdef NDEBUG
#error "Unit tests must be compiled without -DNDEBUG flag"
#endif

#include <assert.h>

#ifdef NDEBUG
#error "Unit tests must be compiled without -DNDEBUG flag"
#endif

//      a-------|
//              |ab--|
//      b--|bb--|    |
//         |         |abc--
//         |c--------|
//
//  t = 0  1    3    5.5     inf
const char *tstInput =
    " # this is a comment\n"
    "time fixed  T0=0\n"
    "time free   Tc=1\n"
    "time free   Tab=3\n"
    "time free   Tabc=5.5\n"
    "twoN free   2Na=100\n"
    "twoN fixed  2Nb=123\n"
    "twoN free   2Nc=213.4\n"
    "twoN fixed  2Nbb=32.1\n"
    "twoN free   2Nab=222\n"
    "twoN fixed  2Nabc=1.2e2\n"
    "mixFrac free Mc=0.02\n"
    "segment a   t=T0     twoN=2Na    samples=1\n"
    "segment b   t=T0     twoN=2Nb    samples=1\n"
    "segment c   t=Tc     twoN=2Nc    samples=1\n"
    "segment bb  t=Tc     twoN=2Nbb\n"
    "segment ab  t=Tab    twoN=2Nab\n"
    "segment abc t=Tabc   twoN=2Nabc\n"
    "mix    b  from bb + Mc * c\n"
    "derive a  from ab\n"
    "derive bb from ab\n"
    "derive ab from abc\n"
    "derive c  from abc\n";
int main(int argc, char **argv) {

    int verbose=0;

    if(argc > 1) {
        if(argc!=2 || 0!=strcmp(argv[1], "-v")) {
            fprintf(stderr,"usage: xgptree [-v]\n");
            exit(EXIT_FAILURE);
        }
        verbose = 1;
    }

    const char *fname = "mktree-tmp.lgo";
    FILE       *fp = fopen(fname, "w");
    fputs(tstInput, fp);
    fclose(fp);

	Bounds   bnd = {
		.lo_twoN = 0.0,
		.hi_twoN = 1e7,
		.lo_t = 0.0,
		.hi_t = HUGE_VAL
	};
    GPTree *g = GPTree_new(fname, bnd);
    GPTree *g2 = GPTree_dup(g);
    assert(GPTree_equals(g, g2));

    const LblNdx lblndx = GPTree_getLblNdx(g);
    if(verbose)
        LblNdx_print(&lblndx, stdout);

    GPTree_free(g);
    GPTree_free(g2);

    unitTstResult("GPTree", "untested");
    return 0;
}
#endif
