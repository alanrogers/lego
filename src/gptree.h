#ifndef ARR_GPTREE_H
#  define ARR_GPTREE_H

#  include "typedefs.h"
#  include <gsl/gsl_rng.h>

#  define POPNAMESIZE 30
#  define MAXSAMP ((int)(8*sizeof(tipId_t)))

struct SampNdx {
    unsigned    n;              // number of samples
    char        lbl[MAXSAMP][POPNAMESIZE];
    PopNode    *node[MAXSAMP];
};

void        Gene_tabulate(Gene * self, BranchTab * bt);
void        Gene_free(Gene * gene);
PopNode    *PopNode_new(double *twoN, double *start, NodeStore *ns);
void        PopNode_addChild(PopNode * parent, PopNode * child);
void        PopNode_mix(PopNode * child, double *mPtr, PopNode * introgressor,
                        PopNode * native);
void        PopNode_newGene(PopNode * self, unsigned ndx);
void        PopNode_addSample(PopNode * self, Gene * gene);
Gene       *PopNode_coalesce(PopNode * self, gsl_rng * rng);
void        PopNode_free(PopNode * self);
void        PopNode_clear(PopNode * self);
void        PopNode_print(FILE * fp, PopNode * self, int indent);
void        PopNode_printShallow(PopNode * self, FILE * fp);
PopNode    *PopNode_root(PopNode * self);
void        PopNode_sanityFromLeaf(PopNode * self, const char *file,
                                   int line);
int         PopNode_nsamples(PopNode * self);

void        SampNdx_init(SampNdx * self);
void        SampNdx_addSamples(SampNdx * self, unsigned nsamples,
							   PopNode * pnode);
void        SampNdx_populateTree(SampNdx * self);
unsigned    SampNdx_size(SampNdx * self);

GPTree     *GPTree_new(const char *fname, Bounds bnd);
void        GPTree_free(GPTree *self);

NodeStore  *NodeStore_new(int len, PopNode v[len]);
void        NodeStore_free(NodeStore *self);
PopNode    *NodeStore_alloc(NodeStore *self);

#endif
