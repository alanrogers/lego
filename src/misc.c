#include "misc.h"
#include "binary.h"
#include "lblndx.h"
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <stdarg.h>
#include <assert.h>
#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>
#ifdef _WIN32
#include <windows.h>
#elif defined(MACOS)
#include <sys/param.h>
#include <sys/sysctl.h>
#else
#include <unistd.h>
#endif
#include <execinfo.h>

#define CALLSTACK_SIZE 128
int         comparePtrs(const void *void_x, const void *void_y);

void dostacktrace(const char *file, int line, FILE * ofp) {
    void       *callstack[CALLSTACK_SIZE];
    int         nsymbols = backtrace(callstack, CALLSTACK_SIZE);

    fprintf(ofp, "backtrace returned %d\n", nsymbols);
    fprintf(ofp, "dostacktrace called from %s:%d:\n", file, line);
    backtrace_symbols_fd(callstack, nsymbols, fileno(ofp));
}

/*
 * Describe an option. For use in "usage" functions.
 */
void tellopt(const char *opt, const char *description) {
    fprintf(stderr, "   %s\n      %s\n", opt, description);
    return;
}

/*
 * An almost platform-independent function that returns the number of
 * CPU cores on the current machine.
 * Source: http://stackoverflow.com/questions/150355/
 * programmatically-find-the-number-of-cores-on-a-machine. If I
 * understand the webpage correctly, this code was written by Dirk-Jan
 * Kroon.
 */
int getNumCores(void) {
#ifdef WIN32
    SYSTEM_INFO sysinfo;

    GetSystemInfo(&sysinfo);
    return sysinfo.dwNumberOfProcessors;
#elif defined(MACOS)
    int         nm[2];
    size_t      len = 4;
    uint32_t    count;

    nm[0] = CTL_HW;
    nm[1] = HW_AVAILCPU;
    sysctl(nm, 2, &count, &len, NULL, 0);

    if(count < 1) {
        nm[1] = HW_NCPU;
        sysctl(nm, 2, &count, &len, NULL, 0);
        if(count < 1) {
            count = 1;
        }
    }
    return count;
#else
    return sysconf(_SC_NPROCESSORS_ONLN);
#endif
}

/* eprintf: print error message and exit */
void eprintf(const char *fmt, ...) {
    va_list     args;

    fflush(stdout);

    va_start(args, fmt);
    vfprintf(stderr, fmt, args);
    va_end(args);

    if(fmt[0] != '\0' && fmt[strlen(fmt) - 1] == ':')
        fprintf(stderr, " %s", strerror(errno));
    fprintf(stderr, "\n");
    exit(EXIT_FAILURE);
}

/*
 * Uniform perturbation on log scale. Log10 of new value in range
 * (log10(x)-w, log10(x)+w)
 */
double perturb_ratio_w(double x, double w, gsl_rng *rng) {
	assert(x >= 0.0);
	if(x == 0.0)
		return x;

    double lgx = log10(x);
    return pow(10.0, gsl_ran_flat(rng, lgx-w, lgx+w));
}

/*
 * Uniform perturbation on log scale. New value in range (0.1*x, 10*x).
 */
double perturb_ratio(double x, gsl_rng *rng) {
	return perturb_ratio_w(x, 1.0, rng);
}

/*
 * Uniform perturbation within interval. New value in range (lo, hi).
 */
long double perturb_interval(long double x, long double lo, long double hi, gsl_rng *rng) {
    if(!(lo <= x)) {
        fflush(stdout);
        fprintf(stderr,"%s:%s:%d: lo=%Lf must be <= x=%Lf\n",
                __FILE__, __func__, __LINE__, lo, x);
        exit(1);
    }
    if(!(x < hi)) {
        fflush(stdout);
        fprintf(stderr,"%s:%s:%d: x=%Lf must be < hi=%Lf\n",
                __FILE__, __func__, __LINE__, x, hi);
        exit(1);
    }
    long double c_hi = hi - 2 * DBL_EPSILON;
    long double c_lo = lo + 2 * DBL_EPSILON;
    assert(c_lo < c_hi);
	long double rval = gsl_ran_flat(rng, c_lo, c_hi);
	assert(rval > c_lo);
	assert(rval < c_hi);
	return rval;
}

void checkmem( /*@null@ */ void *obj, const char *file, int line) {
    if(obj == NULL)
        eprintf("%s:%d: allocation error", file, line);
}

/** Divide lower triangle of tab by nreps */
void normItab(double tab[3][3], unsigned nreps){
    int i, j;
    for(i = 1; i < 3; ++i) {
        for(j = 0; j < i; ++j) {
            tab[i][j] /= nreps;
        }
    }
}

void printItab(double tab[3][3], FILE *fp){
    int i, j;
    double s=0.0;
    fprintf(fp, "%3s", "");
    for(j = 0; j < 2; ++j)
        fprintf(fp, " %8d", j);
    putc('\n', fp);
#if 1
    for(i = 1; i < 3; ++i) {
        fprintf(fp, "%2d:", i);
        for(j = 0; j < i; ++j) {
            fprintf(fp, " %8.4lf", tab[i][j]);
            s += tab[i][j];
        }
        putc('\n', fp);
    }
#else
    for(i = 1; i < 3; ++i) {
        fprintf(fp, "%2d:", i);
        for(j = 0; j < i; ++j) {
            fprintf(fp, " %0.20le", tab[i][j]);
            s += tab[i][j];
        }
        putc('\n', fp);
    }
#endif
    printf("Sum: %lf\n", s);
}

void printEtab(double x10, double x20, double x21, FILE *fp) {
    int j;
    fprintf(fp, "%3s", "");
    for(j = 0; j < 2; ++j)
        fprintf(fp, " %8d", j);
    putc('\n', fp);
    fprintf(fp, "%2d: %8.4lf\n", 1, x10);
    fprintf(fp, "%2d: %8.4lf %8.4lf\n", 2, x20, x21);
    printf("Sum: %lf\n", x10+x20+x21);
}

/** Print a scatter plot in PicTeX format */
void pictex(double *x, double *y, unsigned n, const char *xlbl,
            const char *ylbl, const char *plotheading,
            const char *fname) {
    double maxval=DBL_MIN, minval=DBL_MAX, unit, range;
    double xaxis = 0.2;
    int i;
    FILE *fp = fopen(fname, "w");
    if(fp == NULL)
        eprintf("%s:%s:%d: can't open \"%s\" for output\n",
                __FILE__, __func__, __LINE__,
                fname);
    for(i=0; i<n; ++i) {
        maxval = fmax(maxval, fmax(x[i], y[i]));
        minval = fmin(minval, fmin(x[i], y[i]));
    }
    range = maxval - minval;
    if(minval < 0.0) {
        maxval += 0.02*range;
        minval -= 0.02*range;
    }else{
        minval = 0.0;
        maxval *= 1.02;
    }
    range = maxval - minval;
    fputs("%-*-latex-*-\n", fp);
    fputs("\\mbox{\\beginpicture\n", fp);
    fputs("\\headingtoplotskip 0pt\n", fp);
    fprintf(fp,"%% Horizontal axis: %0.3lf\\textwidth\n", xaxis);
    unit = xaxis / range;
    fprintf(fp,"\\setcoordinatesystem units"
            " <%0.5lf\\textwidth,%0.5lf\\textwidth>\n",
            unit, unit);
    fprintf(fp,"\\setplotarea x from %0.3lf to %0.3lf,"
            " y from %0.3lf to %0.3lf\n",
            minval, maxval, minval, maxval);
    fputs("\\axis bottom label {Predicted}\n", fp);
    fprintf(fp, "  ticks numbered from %0.3lf to %0.3lf by %0.3lf /\n",
            minval, maxval, range);
    fputs("\\axis left label {\\begin{sideways}Simulated\\end{sideways}}\n", fp);
    fprintf(fp, "  ticks numbered from %0.3lf to %0.3lf by %0.3lf /\n",
            minval, maxval, range);
    fprintf(fp, "\\plotheading{%s}\n", plotheading);
    fprintf(fp, "\\plot %0.3lf %0.3lf %0.3lf %0.3lf /\n",
            minval, minval, maxval, maxval);
    fputs("\\multiput {$\\circ$} at\n", fp);
    fprintf(fp, "%%%6s %6s\n", xlbl, ylbl);
    for(i=0; i<n; ++i)
        fprintf(fp, " %6.4lf %6.4lf\n", x[i], y[i]);
    fputs("/\n", fp);
    fputs("\\endpicture}\n", fp);
    fclose(fp);
}

/** Print a scatter plot in PicTeX format */
void pictex_par(double *x, double *y, unsigned n, const char *xlbl,
            const char *ylbl, const char *plotheading,
            const char *fname) {
    double maxval=DBL_MIN, minval=DBL_MAX, unit, range;
    double max_par = DBL_MIN, min_par=DBL_MAX, par_unit, par_range;
    double xaxis = 0.2;
    int i;
    FILE *fp = fopen(fname, "w");
    if(fp == NULL)
        eprintf("%s:%s:%d: can't open \"%s\" for output\n",
                __FILE__, __func__, __LINE__,
                fname);
    for(i=0; i<n; ++i) {
        maxval = fmax(maxval, x[i]);
        minval = fmin(minval, x[i]);
	max_par = fmax(max_par, y[i]);
	min_par = fmin(min_par, y[i]);
    }
    range = maxval - minval;
    par_range = max_par - min_par;
    if(minval < 0.0) {
        maxval += 0.02*range;
        minval -= 0.02*range;
    }else{
        minval = 0.0;
        maxval *= 1.02;
    }
    if(min_par < 0.0) {
        max_par += 0.02*par_range;
        min_par -= 0.02*par_range;
    }else{
        min_par = 0.0;
        max_par *= 1.02;
    }
    range = maxval - minval;
    par_range = max_par - min_par;
    fputs("%-*-latex-*-\n", fp);
    fputs("\\mbox{\\beginpicture\n", fp);
    fputs("\\headingtoplotskip 0pt\n", fp);
    fprintf(fp,"%% Horizontal axis: %0.3lf\\textwidth\n", xaxis);
    unit = xaxis / range;
    par_unit = xaxis / par_range;
    fprintf(fp,"\\setcoordinatesystem units"
            " <%0.5lf\\textwidth,%0.5lf\\textwidth>\n",
            unit, par_unit);
    fprintf(fp,"\\setplotarea x from %0.3lf to %0.3lf,"
            " y from %0.3lf to %0.3lf\n",
            minval, maxval, min_par, max_par);
    fprintf(fp, "\\axis bottom label %s\n", xlbl);
    fprintf(fp, "  ticks numbered from %0.3lf to %0.3lf by %0.3lf /\n",
            minval, maxval, range);
    fprintf(fp, "\\axis left label {%s}\n", ylbl);
    fprintf(fp, "  ticks numbered from %0.3lf to %0.3lf by %0.3lf /\n",
            min_par, max_par, par_range);
    fprintf(fp, "\\plotheading{%s}\n", plotheading);
    /*
    fprintf(fp, "\\plot %0.3lf %0.3lf %0.3lf %0.3lf /\n",
            minval, min_par, maxval, max_par); */
    fputs("\\multiput {$\\circ$} at\n", fp);
    fprintf(fp, "%%%6s %6s\n", xlbl, ylbl);
    for(i=0; i<n; ++i)
        fprintf(fp, " %6.4lf %6.4lf\n", x[i], y[i]);
    fputs("/\n", fp);
    fputs("\\endpicture}\n", fp);
    fclose(fp);
}

/*
 * Vector v must be sorted in ascending order before this function is
 * called.  Function returns index of first element in sorted array
 * that is >= val.  The function assumes without checking that the
 * input is sorted. If val > vec[len-1], the function returns len.
 */
unsigned Dbl_first_geq(double val, unsigned len, double v[len]) {
    register unsigned lo, mid, hi;

    assert(len > 0);
    lo = 0;
    hi = len - 1;
    if(val > v[hi])
        return len;
    while(lo < hi) {
        mid = lo + (hi - lo) / 2;
        if(mid == lo)
            break;
        if(v[mid] < val)
            lo = mid;
        else
            hi = mid;
    }
    if(v[lo] >= val)
        hi = lo;

    assert(hi >= 0);
    assert(hi < len);
    assert(v[hi] >= val);
    assert(hi == 0 || v[hi - 1] < val);

    return hi;
}

void unitTstResult(const char *facility, const char *result) {
    printf("%-26s %s\n", facility, result);
}

/*
 * In string str, count the number of contiguous chunks of characters
 * belonging to set.
 */
int strCountSetChunks(const char *str, const char *sep) {
    int         nchunks = 0, i;

    while(*str != '\0') {
        i = strcspn(str, sep);  /* skip chars not in set */
        str += i;
        i = strspn(str, sep);   /* skip chars in set */
        if(i > 0) {
            ++nchunks;
            str += i;
        }
    }
    return nchunks;
}

/// duplicate memory block
void       *memdup(const void *p, size_t n) {
    void       *q;

    assert(p != NULL);
    assert(n > 0);

    q = malloc(n);
    checkmem(q, __FILE__, __LINE__);
    memcpy(q, p, n);
    return q;
}

/// Compare pointers to pointers to two tipId_t values.
///
/// @param void_x,void_y pointers to pointers to tipId_t values
/// @returns <0, 0, or >0 depending on whether the first arg is <,
/// ==, or > the second.
int comparePtrs(const void *void_x, const void *void_y) {
    tipId_t * const * x = (tipId_t * const *) void_x;
    tipId_t * const * y = (tipId_t * const *) void_y;

    // Major sort is on the number of samples
    // represented in the site pattern. Patterns with
    // fewer samples come first.
    int diff1bits = num1bits(**x) - num1bits(**y);
    if(diff1bits)
        return diff1bits;

    // Reverse order of bits so that low-order bit
    // is most significant. This ensures that the
    // sort order of samples corresponds to the
    // order in which they were listed in the input
    // data.
    unsigned rx = reverseBits(**x);
    unsigned ry = reverseBits(**y);

    return ry - rx;
}

/// Generate a label for site pattern tid. Label goes into
/// buff. Function returns a pointer to buff;
char       *patLbl(size_t n, char buff[n], tipId_t tid, LblNdx * lblndx) {
    int         maxbits = 40;
    int         bit[maxbits];
    int         i, nbits;
    nbits = getBits(tid, maxbits, bit);
    buff[0] = '\0';
    char        lbl[100];
    for(i = 0; i < nbits; ++i) {
        snprintf(lbl, sizeof(lbl), "%s",
                 LblNdx_lbl(lblndx, (unsigned) bit[i]));
        if(strlen(buff) + strlen(lbl) >= n)
            eprintf("%s:%s:%d: buffer overflow\n", __FILE__, __func__,
                    __LINE__);
        strcat(buff, lbl);
        if(i + 1 < nbits && 1 + strlen(buff) < n)
            strcat(buff, ":");
    }
    return buff;
}

/// On entry, pat is an array of n tipId_t values. On return,
/// ord[0] is the index of the first value, ord[1] is that of the 2nd,
/// and so on.
void orderpat(int n, unsigned ord[n], tipId_t pat[n]) {
    tipId_t  *ptr[n];
    int i;
    for(i=0; i < n; ++i)
        ptr[i] = pat+i;
    qsort(ptr, (size_t) n, sizeof(ptr[0]), comparePtrs);
    for(i=0; i<n; ++i)
        ord[i] = ptr[i]-pat;
}

/// Return Kullback-Leibler divergence of o relative to e.
double KLdiverg(int n, const double o[n], const double e[n]) {
    int i;
    double kl=0.0;
    assert(Dbl_near(1.0, sum_double(n, o)));
    assert(Dbl_near(1.0, sum_double(n, e)));
    for(i = 0; i < n; ++i) {
        if(o[i]==0.0 && e[i]==0.0)
            continue;
        double q = o[i];
        double p = e[i];
        kl += p*log(p/q);
    }
    return kl;
}

/// Sum an array of doubles. Unwrapped loop.
double sum_double(int n, const double x[n]) {
    register double rval;
    register int i, m;

    assert(n >= 0);

    rval = 0.0;
    m = n % 5;
    for(i = 0; i < m; ++i)
        rval += x[i];

    for(i = m; i < n; i += 5)
        rval += x[i] + x[i + 1] + x[i + 2] + x[i + 3] + x[i + 4];

    return rval;
}
