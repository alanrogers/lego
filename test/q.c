/**
 * @file q.c
 * @brief Calculate site pattern frequencies under model for Q.
 *
 * @copyright Copyright (c) 2014, Alan R. Rogers 
 * <rogers@anthro.utah.edu>. This file is released under the Internet
 * Systems Consortium License, which can be found in file "LICENSE".
 */

#include "misc.h"
#include "gptree.h"
#include <stdio.h>

void segment(const char *name, double t, double n, int samples);
void mix(const char *ego, const char *ancestor, double m,
         const char *introgressor);
void derive(const char *ego, const char *ancestor);


void segment(const char *name, double t, double n, int samples) {
    printf("segment %3s t=%9.6lf N=%9.6lf", name, t, n);
    if(samples)
        printf(" samples=%d", samples);
    putchar('\n');
}

void mix(const char *ego, const char *ancestor, double m,
         const char *introgressor) {
    printf("mix %s from %9.6lf %s + %9.6lf %s\n", ego, 1-m, ancestor, m,
           introgressor);
}

void derive(const char *ego, const char *ancestor) {
    printf("derive %s from %s\n", ego, ancestor);
}

int main(void) {

    double      twoN0 = 2e4;     // haploid size of ancestral human pop
    double      gen = 29.0;      // generation time
    double      s = twoN0 * gen; // years per time unit


    double mN = 0.02;            // N->EV gene flow
    double mD = 0.01;            // D->V gene flow

    // Time backwards from the present, units of twoN0 gen
    double alpha = 25e3/s;       // Denisovan admixture
    double delta = 55e3/s;       // Neanderthal admixture
    double epsilon = 65e3/s;     // age of older Neanderthal fossil
    double zeta = 110e3/s;       // Y and EV split
    double kappa = 427e3/s;      // N and D split
    double lambda = 658e3/s;     // archaics and moderns split

    // population sizes relative to N0
    double K_X, K_Y, K_D, K_XY, K_ND, K_N;
    K_X = K_Y = K_D = K_XY = K_ND = K_N = 1.0;

#if 1
    // Populations differ in size
    K_X = 2.0;
    K_XY = 2.0;
    K_ND = 0.2;
    K_N = 0.1;
    K_D = 0.1;
#elif 0
    // Force coalescent events into ancestral population
    K_X = K_Y = K_D = K_XY = K_ND = K_N = strtod("Inf", 0);
#endif

    // survival probabilities (s) and their complements (f)
    double sND = survival(lambda - kappa, K_ND);
    double fND = 1.0 - sND;
    double sN  = survival(kappa - delta, K_N);
    double fN  = 1.0 - sN;
    double sXY = survival(lambda - zeta, K_XY);
    double fXY = 1.0 - sXY;

    // Expected fraction of each site pattern
    double Inx = (mN*(1.0-mD)*sN*sND + mD*sND + (1.0 - mN)*(1.0 - mD)*sXY)/3.0;
    double Iny = Inx + mN*(1.0 - mD)*(lambda - delta + (1.0 - K_N)*fN
                                      + (1.0 - K_ND)*sN*fND)
        + mD*(lambda - kappa + (1.0 - K_ND)*fND);
    double Ixy = Inx + (1.0 - mN)*(1.0 - mD)*(lambda - zeta + (1.0 - K_XY)*fXY);

    double sum = Inx+Iny+Ixy;
    Inx /= sum;
    Iny /= sum;
    Ixy /= sum;

    printf("# Q\n");
    printf("# 2N0         : %0.0lf\n", twoN0);
    printf("# t/2N0       : delta=%lf alpha=%lf epsilon=%lf\n", delta, alpha, epsilon);
    printf("#             : kappa=%lf lambda=%lf zeta=%lf\n", kappa, lambda, zeta);
    printf("# Rel pop size: X=%lf Y=%lf N=%lf D=%lf\n", K_X, K_Y, K_N, K_D);
    printf("#             : XY=%lf ND=%lf URHUMAN=%lf\n", K_XY, K_ND, 1.0);
    printf("# admixture->Y: mN=%lf mD=%lf\n", mN, mD);
    printf("# Inx         : %lf\n", Inx);
    printf("# Iny         : %lf\n", Iny);
    printf("# Ixy         : %lf\n", Ixy);

    segment("x", 0, K_X, 1);
    segment("y", 0, K_Y, 1);
    segment("w", alpha, K_Y, 0);
    segment("z", delta, K_Y, 0);
    segment("n", delta, K_N, 1);
    segment("d", alpha, K_D, 1);
    segment("xy", zeta, K_XY,0);
    segment("nd", kappa, K_ND,0);
    segment("a", lambda, 1.0, 0);
    mix("y", "w", mD, "d");
    mix("w", "z", mN, "n");
    derive("x", "xy");
    derive("z", "xy");
    derive("xy", "a");
    derive("n", "nd");
    derive("d", "nd");
    derive("nd", "a");
    
    return 0;
}
