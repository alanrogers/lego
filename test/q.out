# Q
# 2N0         : 20000
# t/2N0       : delta=0.094828 alpha=0.043103 epsilon=0.112069
#             : kappa=0.736207 lambda=1.134483 zeta=0.189655
# Rel pop size: X=2.000000 Y=1.000000 N=0.100000 D=0.100000
#             : XY=2.000000 ND=0.200000 URHUMAN=1.000000
# admixture->Y: mN=0.020000 mD=0.000000
# Inx         : 0.168778
# Iny         : 0.200918
# Ixy         : 0.630304
segment   x t= 0.00 N=    2.00 samples=1
segment   y t= 0.00 N=    1.00 samples=1
segment   z t= 0.09 N=    1.00
segment   n t= 0.09 N=    0.10 samples=1
segment  xy t= 0.19 N=    2.00
segment xyn t= 1.13 N=    1.00
mix y from 0.980000 z + 0.020000 n
derive x from xy
derive z from xy
derive xy from xyn
derive n from xyn
